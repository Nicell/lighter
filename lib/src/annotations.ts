import { annotateLine } from "./annotations.inline";
import { annotateLines } from "./annotations.multiline";
import { Annotation } from "./comments";
import { InlineRange, MultiLineRange } from "./range";

export type Token = {
  content: string;
  style: {
    color?: string;
    fontStyle?: "italic";
    fontWeight?: "bold";
    textDecoration?: "underline" | "line-through";
  };
  scopes?: string[];
};

export type TokenGroup = {
  annotationName: string;
  annotationQuery?: string;
  fromColumn: number;
  toColumn: number;
  tokens: Tokens;
};

export type Tokens = (Token | TokenGroup)[];

export type Line = { lineNumber: number; tokens: Tokens };
export type LineGroup = {
  annotationName: string;
  annotationQuery?: string;
  fromLineNumber: number;
  toLineNumber: number;
  lines: Lines;
};
export type Lines = (Line | LineGroup)[];

export function applyAnnotations(lines: Token[][], annotations: Annotation[]) {
  const { inlineAnnotations, multilineAnnotations } =
    splitAnnotations(annotations);

  let annotatedLines = lines.map((line, lineIndex) => {
    const lineAnnotations = inlineAnnotations.filter(
      (annotation) => annotation.range.lineNumber === lineIndex + 1
    );
    return {
      lineNumber: lineIndex + 1,
      tokens: annotateLine(line, lineAnnotations),
    } as Line;
  });

  return annotateLines(annotatedLines, multilineAnnotations);
}

export type InlineAnnotation = Omit<Annotation, "ranges"> & {
  range: InlineRange;
};

export type MultilineAnnotation = Omit<Annotation, "ranges"> & {
  range: MultiLineRange;
};

function splitAnnotations(annotations: Annotation[]) {
  const inlineAnnotations: InlineAnnotation[] = [];
  const multilineAnnotations: MultilineAnnotation[] = [];

  annotations.forEach((annotation) => {
    annotation.ranges.forEach((range) => {
      if ("lineNumber" in range) {
        inlineAnnotations.push({ ...annotation, range });
      } else {
        multilineAnnotations.push({ ...annotation, range });
      }
    });
  });

  return {
    inlineAnnotations,
    multilineAnnotations,
  };
}
