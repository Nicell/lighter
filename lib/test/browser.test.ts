import { highlight, extractAnnotations } from "../dist/browser.esm.mjs";
import { runTests } from "./highlight.js";
import { runAnnotationTests } from "./annotations.js";

runTests({ highlight });
runAnnotationTests({ extractAnnotations, highlight });
