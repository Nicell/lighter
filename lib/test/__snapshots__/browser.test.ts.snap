// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extract annottations 1`] = `
{
  "annotations": [
    {
      "name": "foo",
      "query": "",
      "ranges": [
        {
          "fromLineNumber": 2,
          "toLineNumber": 2,
        },
      ],
    },
  ],
  "code": "const x = 1;
const y = 2;",
}
`;

exports[`extract annottations 2`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "javascript",
  "lines": [
    {
      "lineNumber": 1,
      "tokens": [
        {
          "content": "const",
          "style": {
            "color": "#569CD6",
          },
        },
        {
          "content": " ",
          "style": {
            "color": "#D4D4D4",
          },
        },
        {
          "content": "x",
          "style": {
            "color": "#4FC1FF",
          },
        },
        {
          "content": " = ",
          "style": {
            "color": "#D4D4D4",
          },
        },
        {
          "content": "1",
          "style": {
            "color": "#B5CEA8",
          },
        },
        {
          "content": ";",
          "style": {
            "color": "#D4D4D4",
          },
        },
      ],
    },
    {
      "annotationName": "foo",
      "annotationQuery": "",
      "fromLineNumber": 2,
      "lines": [
        {
          "lineNumber": 2,
          "tokens": [
            {
              "content": "const",
              "style": {
                "color": "#569CD6",
              },
            },
            {
              "content": " ",
              "style": {
                "color": "#D4D4D4",
              },
            },
            {
              "content": "y",
              "style": {
                "color": "#4FC1FF",
              },
            },
            {
              "content": " = ",
              "style": {
                "color": "#D4D4D4",
              },
            },
            {
              "content": "2",
              "style": {
                "color": "#B5CEA8",
              },
            },
            {
              "content": ";",
              "style": {
                "color": "#D4D4D4",
              },
            },
          ],
        },
      ],
      "toLineNumber": 2,
    },
  ],
}
`;

exports[`extract annottations from text 1`] = `
{
  "annotations": [],
  "code": "const x = 1;
// foo[3:5]
const y = 2;",
}
`;

exports[`extract annottations from text 2`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "text",
  "lines": [
    {
      "lineNumber": 1,
      "tokens": [
        {
          "content": "const x = 1;",
          "style": {},
        },
      ],
    },
    {
      "lineNumber": 2,
      "tokens": [
        {
          "content": "// foo[3:5]",
          "style": {},
        },
      ],
    },
    {
      "lineNumber": 3,
      "tokens": [
        {
          "content": "const y = 2;",
          "style": {},
        },
      ],
    },
  ],
}
`;

exports[`highlight js 1`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x",
        "style": {
          "color": "#9CDCFE",
        },
      },
      {
        "content": " = ",
        "style": {
          "color": "#D4D4D4",
        },
      },
      {
        "content": "1",
        "style": {
          "color": "#B5CEA8",
        },
      },
    ],
  ],
}
`;

exports[`highlight text 1`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "text",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {},
      },
    ],
  ],
}
`;

exports[`highlight with empty theme 1`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#bbbbbb",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#bbbbbb",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {
          "color": "#BBBBBB",
        },
      },
    ],
  ],
}
`;

exports[`highlight with scopes 1`] = `
{
  "colors": {
    "activeTabBackground": "#0d1117",
    "activeTabBorder": "#0d1117",
    "activeTabForeground": "#c9d1d9",
    "activeTabTopBorder": "#f78166",
    "background": "#0d1117",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#23863633",
    "diffInsertedTextBackground": "#2386364d",
    "diffRemovedLineBackground": "#da363333",
    "diffRemovedTextBackground": "#da36334d",
    "editorBackground": "#0d1117",
    "editorGroupHeaderBackground": "#010409",
    "foreground": "#c9d1d9",
    "hoverTabBackground": "#0d1117",
    "hoverTabForeground": "#8b949e",
    "iconForeground": "#8b949e",
    "inactiveTabBackground": "#010409",
    "inactiveTabForeground": "#8b949e",
    "lineNumberForeground": "#6e7681",
    "listHoverBackground": "#6e76811a",
    "listHoverForeground": "#c9d1d9",
    "listSelectionBackground": "#6e768166",
    "listSelectionForeground": "#c9d1d9",
    "selectionBackground": "#264F78",
    "sideBarBackground": "#010409",
    "sideBarBorder": "#30363d",
    "sideBarForeground": "#c9d1d9",
    "tabBorder": "#30363d",
    "tabsBorder": "#30363d",
  },
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x",
        "scopes": [
          "variable.other.readwrite.js",
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": " ",
        "scopes": [
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "=",
        "scopes": [
          "keyword.operator.assignment.js",
          "source.js",
        ],
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": " ",
        "scopes": [
          "source.js",
        ],
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "1",
        "scopes": [
          "constant.numeric.decimal.js",
          "source.js",
        ],
        "style": {
          "color": "#79C0FF",
        },
      },
    ],
  ],
}
`;

exports[`highlight with theme 1`] = `
{
  "colors": {
    "activeTabBackground": "#0d1117",
    "activeTabBorder": "#0d1117",
    "activeTabForeground": "#c9d1d9",
    "activeTabTopBorder": "#f78166",
    "background": "#0d1117",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#23863633",
    "diffInsertedTextBackground": "#2386364d",
    "diffRemovedLineBackground": "#da363333",
    "diffRemovedTextBackground": "#da36334d",
    "editorBackground": "#0d1117",
    "editorGroupHeaderBackground": "#010409",
    "foreground": "#c9d1d9",
    "hoverTabBackground": "#0d1117",
    "hoverTabForeground": "#8b949e",
    "iconForeground": "#8b949e",
    "inactiveTabBackground": "#010409",
    "inactiveTabForeground": "#8b949e",
    "lineNumberForeground": "#6e7681",
    "listHoverBackground": "#6e76811a",
    "listHoverForeground": "#c9d1d9",
    "listSelectionBackground": "#6e768166",
    "listSelectionForeground": "#c9d1d9",
    "selectionBackground": "#264F78",
    "sideBarBackground": "#010409",
    "sideBarBorder": "#30363d",
    "sideBarForeground": "#c9d1d9",
    "tabBorder": "#30363d",
    "tabsBorder": "#30363d",
  },
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "x ",
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "=",
        "style": {
          "color": "#FF7B72",
        },
      },
      {
        "content": " ",
        "style": {
          "color": "#C9D1D9",
        },
      },
      {
        "content": "1",
        "style": {
          "color": "#79C0FF",
        },
      },
    ],
  ],
}
`;

exports[`null code 1`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "javascript",
  "lines": [
    [
      {
        "content": "",
        "style": {
          "color": "#D4D4D4",
        },
      },
    ],
  ],
}
`;

exports[`null lang 1`] = `
{
  "colors": {
    "activeTabBackground": "#1E1E1E",
    "activeTabBorder": "#1E1E1E",
    "activeTabForeground": "#ffffff",
    "activeTabTopBorder": undefined,
    "background": "#1E1E1E",
    "colorScheme": "dark",
    "diffInsertedLineBackground": "#9bb95533",
    "diffInsertedTextBackground": "#9ccc2c33",
    "diffRemovedLineBackground": "#ff000033",
    "diffRemovedTextBackground": "#ff000033",
    "editorBackground": "#1E1E1E",
    "editorGroupHeaderBackground": "#252526",
    "foreground": "#D4D4D4",
    "hoverTabBackground": "#2D2D2D",
    "hoverTabForeground": "#ffffff80",
    "iconForeground": "#C5C5C5",
    "inactiveTabBackground": "#2D2D2D",
    "inactiveTabForeground": "#ffffff80",
    "lineNumberForeground": "#858585",
    "listHoverBackground": "#2A2D2E",
    "listHoverForeground": undefined,
    "listSelectionBackground": "#37373D",
    "listSelectionForeground": undefined,
    "selectionBackground": "#264F78",
    "sideBarBackground": "#252526",
    "sideBarBorder": "#252526",
    "sideBarForeground": "#D4D4D4",
    "tabBorder": "#252526",
    "tabsBorder": undefined,
  },
  "lang": "text",
  "lines": [
    [
      {
        "content": "x = 1",
        "style": {},
      },
    ],
  ],
}
`;
